localstack/aws/handlers/service.py

```diff
@@ -0,0 +1,223 @@
+"""A set of common handlers to parse and route AWS service requests."""
+import logging
+import traceback
+from functools import lru_cache
+from typing import Any, Dict, Optional, Union
+
+from botocore.model import ServiceModel
+
+from localstack import config
+from localstack.http import Response
+
+from ..api import CommonServiceException, RequestContext, ServiceException
+from ..api.core import ServiceOperation
+from ..chain import ExceptionHandler, Handler, HandlerChain
+from ..protocol.parser import RequestParser, create_parser
+from ..protocol.serializer import create_serializer
+from ..protocol.service_router import determine_aws_service_name
+from ..skeleton import Skeleton, create_skeleton
+from ..spec import load_service
+
+LOG = logging.getLogger(__name__)
+
+
+class ServiceNameParser(Handler):
+    """
+    A handler that parses heuristically from the request the AWS service the request is addressed to.
+    """
+
+    def __call__(self, chain: HandlerChain, context: RequestContext, response: Response):
+        service = determine_aws_service_name(context.request)
+
+        if not service:
+            return
+
+        context.service = self.get_service_model(service)
+        headers = context.request.headers
+        headers["x-localstack-tgt-api"] = service
```

Comment:
fair point. tbh i'm not sure we even need this header anymore. i put a TODO in there to remove it. i would try to avoid importing legacy code in the new modules.